casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala

```diff
@@ -477,4 +478,19 @@ object BlockAPI {
           "No action taken since other thread is already processing the block."
         )
     }
+
+  def previewPrivateNames[F[_]: Monad: Log](
+      user: ByteString,
+      timestamp: Long,
+      nameQty: Int
+  ): F[PrivateNamePreviewResponse] = {
+    val seed    = DeployData().withUser(user).withTimestamp(timestamp)
+    val rand    = Blake2b512Random(DeployData.toByteArray(seed))
+    val safeQty = nameQty min 1024
+    val ids = for {
+      _  <- 0 until safeQty
+      id = ByteString.copyFrom(rand.next())
+    } yield id
```

Comment:
Can we use forEach or map here? I think we generally don't want to use List as a monad.

Reply:
@goral09 is there some benefit to writing the code that way? It's not self-explanatory, to me.

Reply:
@dckc there is a rule to use least powerful abstraction. `Functor` (map) is less powerful than `Monad` (flatMap). The other thing is that I treated `rand.next()` an effectful call (which I am not sure it is now). Let's assume for now that it is. Effectful calls should be "suspended" (hence the usage of `Sync[F].delay(…)`). If you run a method `A => F[B]` over the list of `A`s you end up with `List[F[B]]` but what you want is `F[List[B]]` and this is where `traverse` goes into play. Its signature is `List[A] => (A => F[B]) => F[List[B]]`.

Reply:
OK, I see now.

`rand.next()` definitely mutates rand; that's an effect, right? otherwise, `rand.next()` would have to return the same thing every time, right? On the other hand, the state in `rand` doesn't escape this `previewPrivateNames` method, so I don't think sychronization is necessary.